### PR TITLE
Added CLOGLOG to normalizeRegressionResult

### PR DESCRIPTION
**Pull request by ethanyxu**
Hi in my application I needed to use a CLOGLOG transformation to the response to a regular numerical regression PMML, so I added the CLOGCLOG case in normalizeRegressionResult. Note the equation is identical to the one in the normalizeClassificationResult() method, so it should be pretty save.

### CODE REVIEW

**Comment by vruusmann** [comment]
The `cloglog` method is not listed among the permitted methods (`none`, `softmax`, `logit` and `exp`) for regression-type models. See the table "Valid combinations" in the specification of the [RegressionModel element] (http://dmg.org/pmml/v4-2-1/Regression.html).

Perhaps this table is outdated. Could you please tell where did you get your PMML file? What is the name/version of the PMML producer application - is it some well known software?

**Comment by ethanyxu** [comment]
Ah I see.

The pmml I'm using is not a direct export from any language. It's a 'hack'
to do a weibull survival model, which is not supported by PMML yet. The
necessary transformation from weibull to linear regression is precisely
cloglog. The results test out fine after adding cloglog.

I don't see a risk of adding a transformation, but perhaps there's other
considerations?

Thanks!

Ethan

On Thu, Nov 19, 2015 at 12:40 PM, Villu Ruusmann <notifications@github.com>
wrote:

> The cloglog method is not listed among the permitted methods (none,
> softmax, logit and exp) for regression-type models. See the table "Valid
> combinations" in the specification of the RegressionModel element
> <http://dmg.org/pmml/v4-2-1/Regression.html>.
>
> Perhaps this table is outdated. Could you please tell where did you get
> your PMML file? What is the name/version of the PMML producer application -
> is it some well known software?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/jpmml/jpmml-evaluator/pull/16#issuecomment-158131189>.
>



-- 
Yifan "Ethan" Xu, PhD

Data Scientist / Statistician
Explorys, IBM Watson Health

Adjunct Faculty
Department of Epidemiology and Biostatistics
Case Western Reserve University

--------------
Email: ethan.yifanxu@gmail.com
Phone: (607) 760-6817
Url: http://epbiwww.case.edu/~yifan/
--------------

**Comment by vruusmann** [comment]
The main risk is that other PMML consumer software will reject such extended models (just like the JPMML-Evaluator library did in the beginning).

There are actually several ways to achieve your goals using standard constructs. For example, you should define an [`Output` element] (http://dmg.org/pmml/v4-2-1/Output.html), which applies the `cloglog`-equivalent transformation to the predicted value. Like this:
```
<RegressionModel>
  <Output>
    <OutputField name="rawPrediction" feature="predictedValue"/>
    <OutputField name="transformedPrediction" feature="transformedValue">
      <Apply function="-">
        <Constant>1</Constant>
        <Apply function="exp">
          <Apply function="*">
            <Constant>-1</Constant>
            <Apply function="exp">
              <FieldRef field="rawPrediction"/>
            </Apply>
          </Apply>
        </Apply>
      </Apply>
    </OutputField>
  </Output>
</RegressionModel>
```

As a next step, you could isolate the definition of the "cloglog" transformation to a [`DefineFunction` element] (http://dmg.org/pmml/v4-2-1/Functions.html). Like this:
```
<TransformationDictionary>
  <DefineFunction name="cloglog">
    <ParameterField name="x" dataType="double" optype="continuous"/>
    <Apply function="-">
      <!-- Add the rest of the mathematical expression here -->
    </Apply>
  </DefineFunction>
</TransformationDictionary>
```

You can invoke such user-defined functions directly using the `Apply` element. Like this:
```
<Apply function="cloglog">
  <FieldRef field="rawValue"/>
</Apply>
```

**Comment by ethanyxu** [comment]
Thanks! I tried it out and it seems to evaluates the model, but I couldn't figure out how to refer to the "transformedValue". The following seems to return the raw predicts.

    FieldName targetName = evaluator.getTargetField();
    double targetValue = (double) results.get(targetName);

I did attempted to look through the PMML official webpage but with little success. I must be missing something important.

**Comment by vruusmann** [comment]
All the values are are contained in the `result` Map object. You can fetch the values of individual fields by looking them up by name. In the current case, you would have to do the following:
```
Map<FieldName, ?> result = evaluator.evaluate(arguments);
double transformedValue = (double)result.get(FieldName.create("transformedValue"));
```

You can obtain the list of output field names by invoking `Evaluator#getOutputFields()`.

**Comment by vruusmann** [comment]
If you still want the transformed value to be mapped to the `Evaluator#getTargetField()` (and not have any output fields around), then you need to build a [model chain] (http://dmg.org/pmml/v4-2-1/MultipleModels.html). The idea is to have a top-level `MiningModel` element, which contains two `RegressionModel` child elements. The first regression model computes the raw result and the cloglog-transformed result. The second regression model then imports the value of the "transformedValue" field and multiplies it by 1.0. It is basically a no-op, but it allows you to hide the internal complexity of the calculation process from the end user.

```
<MiningModel>
  <Segmentation>
    <Segment>
      <True>
      <RegressionModel>
        <!-- omitted for brevity -->
      </RegressionModel>
    </Segment>
    <Segment>
      <True>
      <RegressionModel>
        <MiningSchema>
          <!-- imports the value of the "transformedValue" output field of the previous model -->
          <MiningField name="transformedValue"/>
        </MiningSchema>
        <RegressionTable intercept="0.0">
          <NumericPredictor name="transformedValue" coefficient="1.0"/>
        </RegressionTable>
      </RegressionModel>
    </Segment>
  </Segmentation>
<MiningModel/>
```

**Comment by ethanyxu** [comment]
Thanks again Villu. The code gives the correct answer after specifying the right field, though there's a little bit of performance decline using this user defined function in xml. Testing it on 1.5 million predictor set:
* Adding CLOGLOG to RegressionModelEvaluator class --> time = 9,345 milliseconds
* Adding the function to pmml xml file --> time = 88,495 millisecond

**Comment by vruusmann** [comment]
I wouldn't call a ten times longer evaluation time "a little performance decline".

If you want to use a native `cloglog` function, then you could re-code your `RegressionModel` element as [`GeneralRegressionModel` element] (http://dmg.org/pmml/v4-2-1/GeneralRegression.html). Scroll down to section "LINK-FUNCTION data type" and you will find a wide variety of applicable link functions.

Looks like your use case is beyond the capabilities of the `RegressionModel` element. So, it would be the right decision to switch over to a more flexible model type. This flexibility adds some overhead, but it's  probably around 20-30%, not 1000% as with user-defined functions.

**Comment by vruusmann** [comment]
If you need little inspiration for your work, then I suggest you to check out the following example PMML documents that contain a regression-type `GeneralRegressionModel` element:

* [KNIME example] (https://github.com/jpmml/jpmml-evaluator/blob/master/pmml-knime/src/test/resources/pmml/GeneralRegressionAuto.pmml)
* [R/Rattle example] (https://github.com/jpmml/jpmml-evaluator/blob/master/pmml-rattle/src/test/resources/pmml/GeneralRegressionAuto.pmml)

**Comment by ethanyxu** [comment]
Great tips. Thanks Villu!


On Thu, Nov 19, 2015 at 4:01 PM, Villu Ruusmann <notifications@github.com>
wrote:

> If you need little inspiration for your work, then I suggest you to check
> out the following example PMML documents that contain a regression-type
> GeneralRegressionModel element:
>
>    - KNIME example
>    <https://github.com/jpmml/jpmml-evaluator/blob/master/pmml-knime/src/test/resources/pmml/GeneralRegressionAuto.pmml>
>    - R/Rattle example
>    <https://github.com/jpmml/jpmml-evaluator/blob/master/pmml-rattle/src/test/resources/pmml/GeneralRegressionAuto.pmml>
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/jpmml/jpmml-evaluator/pull/16#issuecomment-158196516>.
>



-- 
Yifan "Ethan" Xu, PhD

Data Scientist / Statistician
Explorys, IBM Watson Health

Adjunct Faculty
Department of Epidemiology and Biostatistics
Case Western Reserve University

--------------
Email: ethan.yifanxu@gmail.com
Phone: (607) 760-6817
Url: http://epbiwww.case.edu/~yifan/
--------------

**Comment by DougManuel** [comment]
vruusmann, thanks for you comments throughout this thread. 

Just a note to say that we have had similar situations as ethanyxu. Our most recent models are Cox PH models with competing risks in R. Our plan is to create non-competing risk models, export to PMML and then modify that PMML file following the discussions in this thread.  

We have been left with the impression that their has been more emphasis and/or use of PMML for machine learning compared to statistical or regression model.

**Comment by vruusmann** [comment]
Older versions of the JPMML-Evaluator library performed the evaluation of `DefineFunction` elements in a very inefficient manner (involved an unnecessary full JVM class path search operation). This inefficiency was fixed in issue #17.

Today, the cost of evaluating a `DefineFunction` element should be negligible. Therefore, my recommendation is to try to stick with the `RegressionModel` model element, and perform the `cloglog` transformation "manually" using a `DefineFunction` element.

If you're exporting Cox PH models from R, then you'll be getting a `GeneralRegressionModel` model element in the first place. But if there is a need to apply extra post-processing to its prediction, then there is no need to avoid the use of `DefineFunction` elements anymore.
